### PR TITLE
CP-25898 move qemu -netdev option into xenopsd 

### PR DESCRIPTION
**Pull request by lindig**
This series of commits moves the code for the -netdev option from qemu wrapper into xenopsd. Handling this option has one small difficulty: We need to open a network device and pass the file descriptor on the command line. This makes opening the device and creating the command line interdependent. Furthermore, the file descriptor needs to be passed to qemu (wrapper and then binary) and closed once passed. For this, we have to employ the fd mapping feature of Forkexecd: on the command line the filedescriptor is encoded as a UUID, which is replaced by Forkexecd when it creates the new process with the open file descriptor.

* Introduce a new type qemu_args that captures argv and open files
* qemu_argv() constructs command line and opens files, returns qemu_args value
* qemu_argv() uses Tuntap.tap_open() (relies on C stubs) to open device
* Forkhelpers.safe_close_and_exec takes care of mapping UUID/fd mapping. It is called via 
  start_daemin()

### CODE REVIEW

File: xc/stubdom.ml [hunk 1]
```diff
@@ -77,9 +77,10 @@ let build (task: Xenops_task.task_handle) ~xc ~xs ~dm ~store_domid ~console_domi
   let path = Printf.sprintf "/vm/%s/image/dmargs" guest_uuid_str in
   (* Remove any 'pty' references from the arguments: XXX why? *)
   let info' = { info with Device.Dm.serial = None; monitor = None } in
-  let args = Device.Dm.cmdline_of_info ~xs ~dm info' false domid in
-  xs.Xs.write path (String.concat " " args);
```
Comment (mg12) [inline xc/stubdom.ml]: @rossla do you know why qemu args are added to xenstore?

Comment (lindig) [reply xc/stubdom.ml]: This could be wrong and a result of re-basing my code.

File: xc/stubdom.ml [hunk 2]
```diff
@@ -77,9 +77,10 @@ let build (task: Xenops_task.task_handle) ~xc ~xs ~dm ~store_domid ~console_domi
...
-  debug "jjd27: written qemu-dm args into xenstore at %s: [%s]" path (String.concat " " args);
+  let args = Device.Dm.qemu_args ~xs ~dm info' false domid in
+  let argv = args.Device.Dm.argv in
+  xs.Xs.write path (String.concat " " argv);
+  debug "jjd27: written qemu-dm args into xenstore at %s: [%s]" path (String.concat " " argv);
```
Comment (kc284) [inline xc/stubdom.ml]: We shouldn't have user specific indicators (jjd27) in the debug messages.

Comment (lindig) [reply xc/stubdom.ml]: I agree. This code was already there but I can take it out.

File: xc/device.ml [hunk 3]
```diff
@@ -2336,19 +2353,25 @@ module Backend = struct
...
-        common @ misc @ (List.concat nics')
+        let none = ["-net"; "none"] in
+        let add_nic (fds, argv) (mac, bridge, devid) =
+          let ifname          = sprintf "tap%d.%d" domid devid in
+          let uuid, fd as tap = tap_open ifname in
+          let args =
+            [ "-device"; sprintf "rtl8139,netdev=tapnet%d,mac=%s,addr=%d" devid mac (devid + 4)
+            ; "-netdev"; sprintf "tap,id=tapnet%d,fd=%s" devid uuid
+            ] in
+          (tap::fds, args@argv) in
```
Comment (mg12) [inline xc/device.ml]: Looks ok, but as a suggestion to increase readabilitiy, can't we return a type qemu_args here and then use an operator (+) or similar over qemu_args to add this result to common just below?

Comment (lindig) [reply xc/device.ml]: I believe building of the command line requires a refactoring in general. See 

https://github.com/lindig/forkexecd/commit/37d5a22aab257d0387abc8132a00d6bc623fba9f

I could add some comments to explain what is happening here: for each NIC, we introduce command line arguments and open files. These are collected over all NICs in the `fold_left` and in the end need to be added to the already existing command line arguments. 

I don't see how introducing an infix operator would help.

**Comment by lindig** [comment]
* Added comments and improved code structure slightly
* Edited existing debug log message
* This passed a BVT

**Comment by lindig** [comment]
Rebased

File: lib/tuntap.ml [hunk 4]
```diff
@@ -0,0 +1,11 @@
+(*
+ * Copyright (c) Citrix
+ *)
+
+external _tap_open : string -> Unix.file_descr = "stub_tap_open"
+
+let tap_open ifname =
```
Comment (mseri) [inline lib/tuntap.ml]: Would it be useful to provide a `with_tap ifname ~fd:(Unix.file_descr -> 'a)` that also makes sure of closing in a finally block?

File: xc/device.ml [hunk 5]
```diff
@@ -2336,19 +2353,35 @@ module Backend = struct
...
-            [["-net"; "none"]]
-        in
-        common @ misc @ (List.concat nics')
+        (* add_nic is used in a fold: it adds fd and command line args
+         * for a nic to the existing fds and arguments (fds, argv)
+         *)
+        let none = ["-net"; "none"] in
+        let add_nic (fds, argv) (mac, bridge, devid) =
+          let ifname          = sprintf "tap%d.%d" domid devid in
+          let uuid, fd as tap = tap_open ifname in
```
Comment (mseri) [inline xc/device.ml]: `fd` is unused, is that expected? In that case let's ignore it with `_` or `_fd`

File: xc/device.ml [hunk 6]
```diff
@@ -2473,8 +2506,20 @@ module Dm = struct
...
-  let __start (task: Xenops_task.task_handle) ~xs ~dm ?(timeout = !Xenopsd.qemu_dm_ready_timeout) l info domid =
-    debug "Device.Dm.start domid=%d args: [%s]" domid (String.concat " " l);
+
+  type action = Start | Restore | StartVNC
+
+  let __start (task: Xenops_task.task_handle)
+      ~xs ~dm ?(timeout = !Xenopsd.qemu_dm_ready_timeout) action info domid =
+
+    let args = match action with
+      | Start    -> qemu_args ~xs ~dm info false domid
```
Comment (mseri) [inline xc/device.ml]: Please use the boolean as named argument `domid_for_vnc`

Comment (lindig) [reply xc/device.ml]: This is not `domid_for_vnc` but the `restore` parameter which has no label.

Comment (mseri) [reply xc/device.ml]: :P Oh I see, sorry I had misread the function signature

**Comment by lindig** [comment]
I have cleaned up the history at https://github.com/lindig/xenopsd/tree/CP-25898-2/ and would force push this here before merging:

```
b4516092 2018-02-28 CP-25898 remove -netdev option from qemu-wrapper                  ..ian Lindig
96a35c4c 2018-02-28 CP-25898 open /dev/net/tun, pass to qemu                          ..ian Lindig
ccbf66ce 2018-02-28 CP-25898 add module tuntap for /dev/net/tun                       ..ian Lindig
```

**Pull request merged**